test/partition-test/puml.cpp

```diff
@@ -0,0 +1,451 @@
+				sum += nodeweights[i];
+			}
+			for (int i = 0; i < nparts; ++i) {
+				nodeweights[i] /= sum;
+			}
+		}
+		else if (nodeweightsetting == "harmonic") {
+			nodeweights = new double[nparts];
+			double sum = 0;
+			for (int i = 0; i < nparts; ++i) {
+				sum += 1. / (i + 1);
+			}
+			for (int i = 0; i < nparts; ++i) {
+				nodeweights[i] = 1. / (sum * (i + 1));
+			}
+		}
+
+		graph.setVertexWeights(vertexweights, 1);
+		graph.set_edgeWeights(edgeweights);
+		double imbalance = readsetting<double>(settings, "imbalance");
+		int seed = readsetting<int>(settings, "seed");
+
+		PUML::PartitionTarget target;
+		target.setImbalance(imbalance);
+		if (nodeweights == nullptr) {
+			target.setVertexWeightsUniform(nparts);
+		}
+		else {
+			target.setVertexWeights(nparts, nodeweights);
+		}
+
+		logDebug() << "Initial stats" << rank << graph.localVertexCount() << graph.localEdgeCount();
+
+		logInfo(rank) << "Starting the actual partitioning.";
+		struct rusage startm;
+		getrusage(RUSAGE_SELF, &startm);
+		double startt = MPI_Wtime();
+		partitionInvoker->partition(partition.data(), graph, target, seed);
+		MPI_Barrier(comm);
+		double endt = MPI_Wtime();
+		struct rusage endm;
+		getrusage(RUSAGE_SELF, &endm);
+		logInfo(rank) << "Actual partitioning done. Now saving stats";
+
+		// counts weighted and unweighted sizes
+		std::vector<int> partitionSize(nparts*4);
+		std::vector<int> partitionSizeLocal(nparts*4);
+		for (int i = 0; i < puml->cells().size(); ++i) {
+			assert(partition[i] >= 0);
+			assert(partition[i] < nparts);
+			++partitionSizeLocal[partition[i]];
+		}
+		if (vertexweights == nullptr) {
+			for (int i = 0; i < nparts; ++i) {
+				partitionSizeLocal[i + nparts] = partitionSizeLocal[i];
+			}
+		}
+		else {
+			for (int i = 0; i < puml->cells().size(); ++i) {
+				partitionSizeLocal[partition[i] + nparts] += vertexweights[i];
+			}
+		}
+
+		std::vector<int> detail_partition(nparts*nparts);
+		std::vector<int> localDetailPartition(nparts*nparts);
+
+		graph.forallLocalEdges<int>(
+			partition,
+			[&](int fid, int lid, const int& g, const int& l, int id) {
+				if (g != l) {
+					++partitionSizeLocal[g + nparts * 2];
+					if (edgeweights == nullptr) {
+						++partitionSizeLocal[g + nparts * 3];
+						++localDetailPartition[g * nparts + l];
+					}
+					else {
+						partitionSizeLocal[g + nparts * 3] += edgeweights[id];
+						localDetailPartition[g * nparts + l] += edgeweights[id];
+					}
+				}
+			}
+		);
+
+		MPI_Reduce(partitionSizeLocal.data(), partitionSize.data(), partitionSizeLocal.size(), MPI_INT, MPI_SUM, 0, comm);
+		MPI_Reduce(localDetailPartition.data(), detail_partition.data(), localDetailPartition.size(), MPI_INT, MPI_SUM, 0, comm);
+
+		long peakmemstart = startm.ru_maxrss;
+		long peakmemend = endm.ru_maxrss;
+
+		long peakmemlocal[] = {peakmemstart, peakmemend};
+		long peakmem[2];
+		long avgpeakmem[2];
+		MPI_Reduce(peakmemlocal, peakmem, 2, MPI_LONG, MPI_MAX, 0, comm);
```

Comment:
Could use std::array, and MPI_IN_PLACE

Reply:
Yep, at least one of the `MPI_Reduce`s could do that—but to be honest, I'd leave it like that for the test code, at least for now. (reworking the tests properly is a task on its own, most likely)